src/winstanley/wdl.bnf

```diff
@@ -50,13 +50,22 @@ runtime_block ::= RUNTIME map
 parameter_meta_block ::= PARAMETER_META map
 meta_block ::= META map
 
-declaration ::= type_e postfix_quantifier? IDENTIFIER setter?
+declaration ::= type_e IDENTIFIER setter?
 setter ::= EQUAL expression
-postfix_quantifier ::= QMARK | PLUS
 
 map ::= LBRACE kv* RBRACE
 kv ::= IDENTIFIER COLON expression
-type_e ::= TYPE (LSQUARE type_e (COMMA type_e)* RSQUARE)?
+
+type_e ::= array_expression_with_postfix | non_array_expression
+non_array_expression ::= (PRIMITIVE  | map_expression | pair_expression) QMARK?
+array_expression_with_postfix ::= array_expression (QMARK | PLUS)?
+
+no_postfix_quantifier_type ::= array_expression | map_expression | PRIMITIVE | pair_expression
+
+array_expression ::= ARRAY LSQUARE no_postfix_quantifier_type RSQUARE 
+map_expression ::=  MAP_TYPE LSQUARE primitive COMMA no_postfix_quantifier_type RSQUARE
+pair_expression ::=  PAIR LSQUARE no_postfix_quantifier_type COMMA no_postfix_quantifier_type RSQUARE
```

Comment:
That applies to the other types above too right ? Like `Array[Array[String]+]` should be allowed ?